main.cpp

```diff
@@ -261,17 +209,18 @@ Use a service like https://www.diffchecker.com/diff to compare your output.
 #include <functional>
 #include <memory>
 
-struct IntType;
+struct FloatTYpe;
 struct DoubleType;
-struct FloatType; 
-
+struct IntTYpe;
+template<typename T>
+struct Numeric;
```

Comment:
Don't forward declare templated types.  
Move the Point class to be defined AFTER your templated class
- [ ] revise